src/evaluators/source-4-3.js

```diff
@@ -0,0 +1,911 @@
+    return list_ref(stmt, 1);
+}
+
+function cond_expr_cons(stmt) {
+    return list_ref(stmt, 2);
+}
+
+function cond_expr_alt(stmt) {
+    return list_ref(stmt, 3);
+}
+
+function is_function_definition(stmt) {
+    return is_tagged_list(stmt, "function_definition");
+}
+
+function function_definition_parameters(stmt) {
+    return head(tail(stmt));
+}
+
+function function_definition_body(stmt) {
+    return head(tail(tail(stmt)));
+}
+
+function is_return_statement(stmt) {
+    return is_tagged_list(stmt, "return_statement");
+}
+
+function return_statement_expression(stmt) {
+    return head(tail(stmt));
+}
+
+function make_return_value(content) {
+    return list("return_value", content);
+}
+
+function is_return_value(value) {
+    return is_tagged_list(value, "return_value");
+}
+
+function return_value_content(value) {
+    return head(tail(value));
+}
+
+function is_sequence(stmt) {
+    return is_tagged_list(stmt, "sequence");
+}
+
+function make_sequence(stmts) {
+    return list("sequence", stmts);
+}
+
+function sequence_statements(stmt) {
+    return head(tail(stmt));
+}
+
+function is_empty_sequence(stmts) {
+    return is_null(stmts);
+}
+
+function is_last_statement(stmts) {
+    return is_null(tail(stmts));
+}
+
+function first_statement(stmts) {
+    return head(stmts);
+}
+
+function rest_statements(stmts) {
+    return tail(stmts);
+}
+
+function sequence_actions(stmts) {
+    return head(tail(stmts));
+}
+
+function local_names(stmt) {
+    if (is_sequence(stmt)) {
+        const stmts = sequence_statements(stmt);
+        return is_empty_sequence(stmts) ?
+            null :
+            insert_all(
+                local_names(first_statement(stmts)),
+                local_names(make_sequence(
+                    rest_statements(stmts))));
+    } else {
+        return is_constant_declaration(stmt) ?
+            list(constant_declaration_name(stmt)) :
+            is_variable_declaration(stmt) ?
+            list(variable_declaration_name(stmt)) :
+            null;
+    }
+}
+
+function insert_all(xs, ys) {
+    return is_null(xs) ?
+        ys :
+        is_null(member(head(xs), ys)) ?
+        pair(head(xs), insert_all(tail(xs), ys)) :
+        error(head(xs), "multiple declarations of: ");
+}
+
+function is_block(stmt) {
+    return is_tagged_list(stmt, "block");
+}
+
+function make_block(stmt) {
+    return list("block", stmt);
+}
+
+function block_body(stmt) {
+    return head(tail(stmt));
+}
+
+function is_application(stmt) {
+    return is_tagged_list(stmt, "application");
+}
+
+function operator(stmt) {
+    return head(tail(stmt));
+}
+
+function operands(stmt) {
+    return head(tail(tail(stmt)));
+}
+
+function no_operands(ops) {
+    return is_null(ops);
+}
+
+function first_operand(ops) {
+    return head(ops);
+}
+
+function rest_operands(ops) {
+    return tail(ops);
+}
+
+function make_primitive_function(impl) {
+    return list("primitive", impl);
+}
+
+
+function make_function(impl) {
```

Comment:
A minor clarification: this function does not seem to be used anywhere, is there a purpose for it?

Reply:
Thank you! Updated at 547be0b